contracts/Strategy.sol

```diff
@@ -53,80 +222,624 @@ contract Strategy is BaseStrategy {
+            }
+            if (amountToFree > 0) {
+                uint256 sharesNeeded = wantToBentoShares(amountToFree, true);
+                uint256 bentoShares = sharesInBento();
+
+                uint256 sharesToFreeFromKashi =
+                    bentoShares <= sharesNeeded
+                        ? sharesNeeded.sub(bentoShares)
+                        : 0;
+
+                uint256 sharesFreedFromKashi = 0;
+
+                // Find the lowest apr pair with at least the lesser of
+                //   - the amount to free
+                //   - the mean assets per pair
+                uint256 lowestInterestIndex =
+                    lowestInterestPairIndex(
+                        Math.min(
+                            sharesToFreeFromKashi,
+                            wantToBentoShares(
+                                estimatedTotalAssets().div(kashiPairs.length),
+                                true
+                            )
+                        )
+                    );
+                if (lowestInterestIndex != type(uint256).max) {
+                    sharesFreedFromKashi = liquidateKashiPair(
+                        lowestInterestIndex,
+                        sharesToFreeFromKashi
+                    );
+                }
 
-        uint256 totalAssets = want.balanceOf(address(this));
-        if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+                for (
+                    uint256 i = 0;
+                    i < kashiPairs.length &&
+                        sharesToFreeFromKashi > sharesFreedFromKashi;
+                    i++
+                ) {
+                    if (i == lowestInterestIndex) continue; // we already visited this
+
+                    sharesFreedFromKashi = sharesFreedFromKashi.add(
+                        liquidateKashiPair(
+                            i,
+                            sharesToFreeFromKashi.sub(sharesFreedFromKashi)
+                        )
+                    );
+                }
+
+                bentoBox.withdraw(
+                    BIERC20(address(want)),
+                    address(this),
+                    address(this),
+                    0,
+                    sharesInBento()
+                );
+            }
+
+            _liquidatedAmount = balanceOfWant();
+
+            if (_amountNeeded > _liquidatedAmount) {
+                _loss = _amountNeeded.sub(_liquidatedAmount);
+            }
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
-    function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 _liquidatedAmount)
+    {
+        (_liquidatedAmount, ) = liquidatePosition(type(uint256).max);
```

Comment:
should this be instead `estimatedTotalAssets().sub(wantBalance)` instead of uint max?

since in line 295 we are effectively forcing the value to that? is there any issue with calculating the actual total here?

Reply:
✅